mempool/v1/mempool.go

```diff
@@ -129,7 +127,18 @@ func (txmp *TxMempool) SizeBytes() int64 { return atomic.LoadInt64(&txmp.txsByte
 //
 // The caller must hold an exclusive mempool lock (by calling txmp.Lock) before
 // calling FlushAppConn.
-func (txmp *TxMempool) FlushAppConn() error { return txmp.proxyAppConn.FlushSync() }
+func (txmp *TxMempool) FlushAppConn() error {
+	// N.B.: We have to issue the call outside the lock so that its callback can
+	// fire.  It's safe to do this, the flush will block until complete.
+	//
+	// We could just not require the caller to hold the lock at all, but the
+	// semantics of the Mempool interface require the caller to hold it, and we
+	// can't change that without disrupting existing use.
+	txmp.mtx.Unlock()
+	defer txmp.mtx.Lock()
```

Comment:
I just checked and this is coming from `v0.34.x` (committed after v0.34.20), I wasn't aware of this change. I will discuss this internally in Tendermint as well. Here's my understanding why I am not convinced we should release the lock here.

The main rationale for holding the lock here is that, otherwise, p2p might deliver new TXs received from the network. If the mempool is not locked, my understanding is that those newly arrived transactions would trigger `CheckTx`, so we couldn't guarantee that _nothing_ is going on at the application level when we call commit.

In any case, the behavior of this should be consistent with the `v0` mempool.